src/hyperlight_host/src/hypervisor/mod.rs

```diff
@@ -314,201 +557,93 @@ pub(crate) trait Hypervisor: Debug + Send {
     let region = mem_regions.find(|region| region.guest_region.contains(&gpa));
 
-    if let Some(region) = region
-        && (!region.flags.contains(access_info)
-            || region.flags.contains(MemoryRegionFlags::STACK_GUARD))
-    {
-        return Some(HyperlightExit::AccessViolation(
-            gpa as u64,
-            access_info,
-            region.flags,
-        ));
+    if let Some(region) = region {
+        if region.region_type == MemoryRegionType::GuardPage {
+            return Some(MemoryAccess::StackGuardPageViolation);
+        } else if !region.flags.contains(tried) {
+            return Some(MemoryAccess::AccessViolation(region.flags));
+        }
     }
     None
 }
 
-/// A virtual CPU that can be run until an exit occurs
-pub struct VirtualCPU {}
-
-impl VirtualCPU {
-    /// Run the given hypervisor until a halt instruction is reached
-    #[instrument(err(Debug), skip_all, parent = Span::current(), level = "Trace")]
-    pub(crate) fn run(
-        hv: &mut dyn Hypervisor,
-        #[cfg(gdb)] dbg_mem_access_fn: Arc<Mutex<SandboxMemoryManager<HostSharedMemory>>>,
-    ) -> Result<()> {
-        // Keeps the trace context and open spans
-        #[cfg(feature = "trace_guest")]
-        let mut tc = crate::sandbox::trace::TraceContext::new();
-
-        loop {
-            #[cfg(feature = "trace_guest")]
-            let result = {
-                let result = hv.run(&mut tc);
-                // End current host trace by closing the current span that captures traces
-                // happening when a guest exits and re-enters.
-                tc.end_host_trace();
-
-                // Handle the guest trace data if any
-                if let Err(e) = hv.handle_trace(&mut tc) {
-                    // If no trace data is available, we just log a message and continue
-                    // Is this the right thing to do?
-                    log::debug!("Error handling guest trace: {:?}", e);
-                }
-
-                result
-            };
-            #[cfg(not(feature = "trace_guest"))]
-            let result = hv.run();
-
-            match result {
-                #[cfg(gdb)]
-                Ok(HyperlightExit::Debug(stop_reason)) => {
-                    if let Err(e) = hv.handle_debug(dbg_mem_access_fn.clone(), stop_reason) {
-                        log_then_return!(e);
-                    }
-                }
-
-                Ok(HyperlightExit::Halt()) => {
-                    break;
-                }
-                Ok(HyperlightExit::IoOut(port, data, rip, instruction_length)) => {
-                    hv.handle_io(port, data, rip, instruction_length)?
-                }
-                Ok(HyperlightExit::Mmio(addr)) => {
-                    #[cfg(crashdump)]
-                    crashdump::generate_crashdump(hv)?;
+/// A trait for platform-specific interrupt handle implementation details
+pub(crate) trait InterruptHandleImpl: InterruptHandle {
```

Comment:
Maybe I don't understand it well, but what does this second trait help with?
I see both Linus and Windows variants implement it and `InterruptHandle` at the same time, no default implementation. So it is as if it is a single trait split into two smaller traits, right?

Would this help with next PRs?

Reply:
Got it. It makes sense